source/fuzz/fact_manager.cpp

```diff
@@ -588,10 +588,29 @@ void FactManager::DataSynonymAndIdEquationFacts::AddEquationFactRecursive(
   // Now try to work out corollaries implied by the new equation and existing
   // facts.
   switch (opcode) {
+    case SpvOpConvertSToF:
+    case SpvOpConvertUToF: {
+      // Equation form: a = float(b)
+      for (const auto& fact : id_equations_) {
+        for (const auto& equation : fact.second) {
```

Comment:
I don't get why this way round isn't symmetric to the code in your previous commit for SPvOpConvertFTo[S|U] - i.e., why do you have a pair of for loops here but just a single for loop below?  If they really need to be different then please explain with a comment.

Reply:
Sure ;). `GetEquations` method returns all equations for some particular id (i.e. given a descriptor of an object, it returns a set of opcodes and operands that define those equations - this method internally uses a map from the left-hand side of the equation to all known right-hand sides). 

This works well in the initial case. That being said, if we know that `a = int(b)` (or more specifically `%a = OpConvertFTo[S|U] %b`) then we only need to find an equation `b = float(c)` (`%b = OpConvert[S|U]ToF %c`). The crucial thing to note here is that `b` appears on the left-hand side of the equation. Thus, we can easily use `GetEquations` method.

However, if we know that `a = float(b)` (`%a = OpConvert[S|U]ToF %b`), we need to find an equation of the form `c = int(a)` (`%c = OpConvertFTo[S|U] %a`). We can't use `GetEquations` in this case since we search by `a` which appears on the right-hand side of the equation. The simplest solution is to iterate over all equations and check for each one that its right-hand side matches `a`. That's what we do.

Reply:
Fixed in d6ecf7e.

Reply:
I didn't look at your fix yet, but I would have done it like this:

```
    case SpvOpConvertFToS:
    case SpvOpConvertFToU: {
      // Equation form: "a = int(b)"
      for (const auto& equation : GetEquations(rhs_dds[0])) {
        if (equation.opcode == SpvOpConvertSToF || equation.opcode == SpvOpConvertUToF) {
          // Equation form: "b = float(c)"
          // We can thus infer "a = c"
          AddDataSynonymFactRecursive(lhs_dd, *equation.operands[0], context);
        }
      }
      break;
    }
    case SpvOpConvertSToF:
    case SpvOpConvertUToF: {
      // Equation form: "a = float(b)"
      for (const auto& equation : GetEquations(rhs_dds[0])) {
        if (equation.opcode == SpvOpConvertFToS || equation.opcode == SpvOpConvertFToU) {
          // Equation form: "b = int(c)"
          // We can thus infer "a = c"
          AddDataSynonymFactRecursive(lhs_dd, *equation.operands[0], context);
        }
      }
      break;
    }
```

However, I've just realised something bad: these convert instructions don't do a bitcast (I had gotten confused and thought that they did!) - so these equation facts are not valid.  We need to re-think this one; I'll message you.

Reply:
Oh, I guess you already knew this and that's why you did it the asymmetric way - that makes sense I think, but let me think about it a bit more.